actors/migration/nv15/miner.go

```diff
@@ -3,79 +3,201 @@ package nv15
-		return nil, err
+		return nil, xerrors.Errorf("getting inState: %w", err)
 	}
 
-	sectorsOut, err := migrateSectors(ctx, store, inState.Sectors)
+	outState := fromv6State(inState)
+	ctxStore := adt.WrapStore(ctx, store)
+
+	sectorsOut, err := migrateSectors(ctx, ctxStore, in.cache, in.address, inState.Sectors)
 	if err != nil {
-		return nil, err
+		return nil, xerrors.Errorf("failed to migrate sectors: %w", err)
 	}
 
-	outState := miner7.State{
-		Info:                       inState.Info,
-		PreCommitDeposits:          inState.PreCommitDeposits,
-		LockedFunds:                inState.LockedFunds,
-		VestingFunds:               inState.VestingFunds,
-		FeeDebt:                    inState.FeeDebt,
-		InitialPledge:              inState.InitialPledge,
-		PreCommittedSectors:        inState.PreCommittedSectors,
-		PreCommittedSectorsCleanUp: inState.PreCommittedSectorsCleanUp,
-		AllocatedSectors:           inState.AllocatedSectors,
-		Sectors:                    sectorsOut,
-		ProvingPeriodStart:         inState.ProvingPeriodStart,
-		CurrentDeadline:            inState.CurrentDeadline,
-		Deadlines:                  inState.Deadlines,
-		EarlyTerminations:          inState.EarlyTerminations,
-		DeadlineCronActive:         inState.DeadlineCronActive,
+	outState.Sectors = sectorsOut
+
+	deadlinesOut, err := migrateDeadlines(ctx, ctxStore, in.cache, inState.Deadlines, sectorsOut)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to migrate deadlines: %w", err)
 	}
 
+	outState.Deadlines = deadlinesOut
+
 	newHead, err := store.Put(ctx, &outState)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to flush outState: %w", err)
+	}
+
 	return &actorMigrationResult{
 		newCodeCID: m.migratedCodeCID(),
 		newHead:    newHead,
-	}, err
+	}, nil
 }
 
 func (m minerMigrator) migratedCodeCID() cid.Cid {
 	return builtin7.StorageMinerActorCodeID
 }
 
-func migrateSectors(ctx context.Context, store cbor.IpldStore, inRoot cid.Cid) (cid.Cid, error) {
-	ctxStore := adt.WrapStore(ctx, store)
-	inArray, err := adt.AsArray(ctxStore, inRoot, miner6.SectorsAmtBitwidth)
+func migrateSectors(ctx context.Context, store adt.Store, cache MigrationCache, minerAddr address.Address, inRoot cid.Cid) (cid.Cid, error) {
+	return cache.Load(SectorsAmtKey(inRoot), func() (cid.Cid, error) {
+		inArray, err := adt.AsArray(store, inRoot, miner6.SectorsAmtBitwidth)
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to read sectors array: %w", err)
+		}
+
+		okIn, prevInRoot, err := cache.Read(MinerPrevSectorsInKey(minerAddr))
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to get previous inRoot from cache: %w", err)
+		}
+
+		okOut, prevOutRoot, err := cache.Read(MinerPrevSectorsOutKey(minerAddr))
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to get previous outRoot from cache: %w", err)
+		}
+
+		if okIn != okOut {
+			return cid.Undef, xerrors.Errorf("cannot have only one of the inRoot and outRoot in the cache")
+		}
```

Comment:
If we get an error here, we may want to just log it, assume _nothing_ is cached, and do the migration.

Reply:
Yeah, this was a convo between Zen and me upstream^, neither of us felt too strongly...I think I'll adopt your suggestion